firebase-firestore/src/main/java/com/google/firebase/firestore/LoadBundleTask.java

```diff
@@ -484,47 +493,54 @@ public LoadBundleTask addOnProgressListener(
   private void removeOnProgressListener(
       @NonNull OnProgressListener<LoadBundleTaskProgress> listener) {
     synchronized (lock) {
-      progressListenerQueue.remove(new ManagedListener(/* executor= */ null, listener));
+      progressListeners.remove(new ManagedListener(/* executor= */ null, listener));
     }
   }
 
   void setResult(@Nullable LoadBundleTaskProgress result) {
+    Queue<ManagedListener> currentListeners;
     synchronized (lock) {
       snapshot = result;
-      for (ManagedListener progressListener : progressListenerQueue) {
-        progressListener.maybeRun(result);
-      }
-      progressListenerQueue.clear();
+      currentListeners = new ArrayDeque<>(progressListeners);
+      progressListeners.clear();
     }
+    invokeListeners(currentListeners, result);
     completionSource.setResult(result);
   }
 
   void setException(@NonNull Exception exception) {
+    Queue<ManagedListener> currentListeners;
+    LoadBundleTaskProgress snapshot;
     synchronized (lock) {
-      LoadBundleTaskProgress lastSnapshot =
-          snapshot != null ? snapshot : LoadBundleTaskProgress.INITIAL;
       snapshot =
           new LoadBundleTaskProgress(
-              lastSnapshot.getDocumentsLoaded(),
-              lastSnapshot.getTotalDocuments(),
-              lastSnapshot.getBytesLoaded(),
-              lastSnapshot.getTotalBytes(),
+              this.snapshot.getDocumentsLoaded(),
+              this.snapshot.getTotalDocuments(),
+              this.snapshot.getBytesLoaded(),
+              this.snapshot.getTotalBytes(),
               exception,
               LoadBundleTaskProgress.TaskState.ERROR);
-      for (ManagedListener progressListener : progressListenerQueue) {
-        progressListener.maybeRun(this.snapshot);
-      }
-      progressListenerQueue.clear();
+      this.snapshot = snapshot;
+
+      currentListeners = new ArrayDeque<>(progressListeners);
+      progressListeners.clear();
     }
+    invokeListeners(currentListeners, snapshot);
     completionSource.setException(exception);
   }
 
-  void updateProgress(LoadBundleTaskProgress snapshot) {
+  void updateProgress(LoadBundleTaskProgress progressUpdate) {
+    Queue<ManagedListener> currentListeners;
```

Comment:
I thought about this more: Since I am always invoking the callback on an executor, there should only be two scenarios:

- `updateProgress` is executed on the same thread as the callbacks. This means that all `execute` calls will only run after the loop has finished processing.
- `updateProgress` is executed on a different thread from the callbacks. This means that the other thread won't be able to grab the lock to modify the progress listener.

With that, I am reasonably confident that we don't need the extra copy and can simplify this quite a bit.

(FYI - I could implement your suggestion above by changing the data structure to our internal ImmutableSet/Map implementation, but I do still prefer to keep in sync with Android's Task API.

Reply:
Ahh yes that makes sense. I missed the fact that the callbacks were being enqueued on an executor. In that case, I rescind my comment about the ConcurrentModificationException.